language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,370 @@
+               -> m [(Name, Normalized (Type TyName uni ()))]
+bindingNewVars = unNormalized >>> \case
+  -- @effectfully: is it needed to rename the termbind's normalized-type? I don't think so.
+  TermBind _ _ (VarDecl _ n ty) _rhs ->
+      pure [(n, Normalized $ void ty)]
+
+  DatatypeBind _ dt@(Datatype ann _ tyargs desName vdecls) -> do
+      -- build the patternfunctor's "normalized" type and from it the "normalized" destructor
+      pf <- mkDatatypePatternFunctor ann dt
+
+      let -- build the normalized constructors
+          constrs = fmap (\(VarDecl _ constrName constrTy) -> (constrName, PIR.mkIterTyForall tyargs constrTy))
+                         vdecls
+          des = (desName, mkDestructorTy ann pf dt)
+          newStructors = des:constrs
+
+      -- rename all *-structors before we put them into the env, to make sure we preserve global uniqueness
+      forMOf (traversed._2) newStructors $
+          Normalized . void <<$>> rename
+
+  _ -> pure [] -- typebinds don't introduce vardecls
+
+-- Helpers
+----------
+
+-- | A wrapper for safety: check a type to be *-kinded, and only then normalize it
+-- NB: Output is normalized (but not wrapped in Normalized because it is used inside a Binding (and binding is not a functor))
+checkStarNorm :: AsTypeError e term uni ann
+              => Type TyName uni ann
+              -> TypeCheckM uni e (Type TyName uni ann)
+checkStarNorm ty = unNormalized <$>
+                   normalizeTypeM ty <*
+                   checkKindM (typeAnn ty) ty (Type ())
+
+-- | Scope a typechecking computation with the given binding's newly-introducing type (if there is one)
+withNewType :: Binding TyName name uni () -> TypeCheckM uni e a -> TypeCheckM uni e a
+withNewType = \case
+    TypeBind _ (TyVarDecl _ tn k) _ -> withTyVar tn k
+    DatatypeBind _ (Datatype _ (TyVarDecl _ tn k) _ _ _) -> withTyVar tn k
+    _ -> id -- no type to introduce
+
+-- | warning: Does not no do conflict check
+withNewTypes :: Foldable f => f (Binding TyName name uni ()) -> TypeCheckM uni e a -> TypeCheckM uni e a
+withNewTypes = flip $ foldr withNewType
+
+-- | warning: Does not no do conflict check
+withTyVarDecls :: [TyVarDecl TyName ann] -> TypeCheckM uni e a -> TypeCheckM uni e a
+withTyVarDecls = flip . foldr $ \(TyVarDecl _ n k) -> withTyVar n $ void k
+
+-- | warning: Does not no do conflict check
+withVars :: [(Name, Normalized (Type TyName uni ()))] -> TypeCheckM uni e a -> TypeCheckM uni e a
+withVars = flip . foldr $ uncurry withVar
+
+-- |  returns ([[TypeCon tyarg1] ... tyargn])
+mkAppliedTyCon :: Datatype TyName Name uni ann -> Type TyName uni ann
+mkAppliedTyCon (Datatype ann tycon tyargs _ _) =
+    PIR.mkIterTyApp ann (PIR.mkTyVar ann tycon) $ PIR.mkTyVar ann <$> tyargs
+
+-- Turned off duplicate/recconflict checks
+-- hasDuplicate :: Ord a => [a] -> Maybe a
+-- hasDuplicate l = head <$> (find (\g -> length g > 1) .  group $ sort l)
+-- assertNoDuplicate :: (AsTypeErrorExt e uni ann)
```

Comment:
Please keep code that we don't need somewhere else, there's not much point in committing it to master.